crypto/cms/cms_smime.c

```diff
@@ -854,18 +854,18 @@ int CMS_decrypt(CMS_ContentInfo *cms, EVP_PKEY *pk, X509 *cert,
     }
     if (dcont == NULL && !check_content(cms))
         return 0;
-    if (flags & CMS_DEBUG_DECRYPT)
-        ossl_cms_get0_env_enc_content(cms)->debug = 1;
-    else
-        ossl_cms_get0_env_enc_content(cms)->debug = 0;
-    if (cert == NULL)
-        ossl_cms_get0_env_enc_content(cms)->havenocert = 1;
-    else
-        ossl_cms_get0_env_enc_content(cms)->havenocert = 0;
+    ec = ossl_cms_get0_env_enc_content(cms);
+    ec->debug = (flags & CMS_DEBUG_DECRYPT) != 0;
+    ec->havenocert = cert == NULL;
     if (pk == NULL && cert == NULL && dcont == NULL && out == NULL)
         return 1;
-    if (pk != NULL && !CMS_decrypt_set1_pkey(cms, pk, cert))
+    if (pk != NULL && !CMS_decrypt_set1_pkey(cms, pk, cert)) {
+        /* Prevent mem leak if CMS_decrypt_set1_password() was used before */
+        OPENSSL_clear_free(ec->key, ec->keylen);
```

Comment:
This doesnt look nice to me, so not sure I understand this.. If the setup of ec->key came from calling CMS_decrypt_set1_pkey() (via ossl_cms_RecipientInfo_pwri_crypt()) then why is it freeing it here? Wouldn't the internals of CMS_decrypt_set1_pkey() do this, down in the bowels of the call?

Reply:
I don't understand the details here myself.
There was a mem leak, see the OP, so certainly some fix is needed.
It was pretty tricky for me to find out how to fix it, and it did plug the leak at least in the situation where I encountered it.

Reply:
If you can repo the memleak, then maybe try putting this code in the 'correct' place.. which seems to be in cms_asn1.c

ASN1_NDEF_SEQUENCE(CMS_EncryptedContentInfo) would need a callback.. (See cms_ri_cb() as an example).

i.e. These structures are all ASN1 related objects that have a messed up way of freeing.

Reply:
Meanwhile I've reproduced the mem leak, and understand what goes wrong here:

During a successful `CMS_decrypt_set1_password()` (or earlier `CMS_decrypt_set1_pkey()`) call, `ossl_cms_RecipientInfo_pwri_crypt()` records the derived decryption key in `ec->key` (and  `ec->keylen`).

When then `CMS_decrypt()` calls `CMS_decrypt_set1_pkey()`, which is supposed to set the decryption key as well, it can happen that this call fails (for instance because there is no matching recipient), but then `ec->key` is still set, and `CMS_decrypt()` exits without deallocating it (unless the workaround I had placed here frees the key).

Usually, the key is implicitly deallocated a little later in `CMS_decrypt()`, namely by `CMS_dataInit()` which frees the key at the end of `ossl_cms_EncryptedContent_init_bio()` -
what a weird design.

I've improved my fix by moving the deallocation of the key into `CMS_decrypt_set1_pkey(_and_peer)`, which now frees any previously recorded key beforehand.
I've also extended the documentation to describe the key memory management by these functions as far as relevant to users.

Reply:
Nice detective work.